internal/configsource/vaultconfigsource/session.go

```diff
@@ -0,0 +1,318 @@
+	originalVersion := v.extractVersionMetadata(mdMap, "created_time", "version")
+	if originalVersion == nil {
+		v.logger.Warn("Failed to extract version metadata to create to create polling watcher for vault config source", zap.String("path", v.path))
+		return watcherNotSupported, nil
+	}
+
+	watcherFn := func() error {
+		metadataPath := strings.Replace(v.path, "/data/", "/metadata/", 1)
+		ticker := time.NewTicker(v.pollInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				metadataSecret, err := v.client.Logical().Read(metadataPath)
+				if err != nil {
+					// Docs are not clear about how to differentiate between temporary and permanent errors.
+					// Assume that the configuration needs to be re-fetched.
+					return fmt.Errorf("failed to read secret metadata at %q: %w", metadataPath, err)
+				}
+
+				if metadataSecret == nil || metadataSecret.Data == nil {
+					return fmt.Errorf("no secret metadata found at %q", metadataPath)
+				}
+
+				const timestampKey = "updated_time"
+				const versionKey = "current_version"
+				latestVersion := v.extractVersionMetadata(metadataSecret.Data, timestampKey, versionKey)
+				if latestVersion == nil {
+					return fmt.Errorf("secret metadata is not in the expected format for keys %q and %q", timestampKey, versionKey)
+				}
+
+				// Per SmartAgent code this is enough to trigger an update but it is also possible to check if the
+				// the valued of the retrieved keys was changed. The current criteria may trigger updates even for
+				// addition of new keys to the secret.
+				if originalVersion.Timestamp != latestVersion.Timestamp || originalVersion.Version != latestVersion.Version {
+					return configsource.ErrValueUpdated
+				}
+			case <-v.doneCh:
+				return configsource.ErrSessionClosed
+			}
+		}
+	}
+
+	return watcherFn, nil
+}
+
+type versionMetadata struct {
+	Timestamp string
+	Version   int64
+}
+
+func (v *vaultSession) extractVersionMetadata(metadataMap map[string]interface{}, timestampKey, versionKey string) *versionMetadata {
+	timestamp, ok := metadataMap[timestampKey].(string)
+	if !ok {
+		v.logger.Warn("Missing or unexpected type for timestamp on the metadata map", zap.String("key", timestampKey))
+		return nil
+	}
+
+	versionNumber, ok := metadataMap[versionKey].(json.Number)
+	if !ok {
+		v.logger.Warn("Missing or unexpected type for version on the metadata map", zap.String("key", versionKey))
+		return nil
+	}
+
+	versionInt, err := versionNumber.Int64()
+	if err != nil {
+		v.logger.Warn("Failed to parse version number into an integer", zap.String("key", versionKey), zap.String("version_number", string(versionNumber)))
+		return nil
+	}
+
+	return &versionMetadata{
+		Timestamp: timestamp,
+		Version:   versionInt,
+	}
+}
+
+// Allows key to be dot-delimited to traverse nested maps.
+func traverseToKey(data map[string]interface{}, key string) interface{} {
+	parts := strings.Split(key, ".")
+
+	for i := 0; ; i++ {
+		partVal := data[parts[i]]
```

Comment:
`parts` may be empty, right?

Reply:
Because the `sep` parameter is never empty it parts always have at least one element. I will add a comment to make it clear.

Reply:
Added comment.